File: Dante2/Module2.07/Zadanie1.04/main.c
#include <stdio.h>
#include "for_loop.h"

int main()
{
    int error;
    void (*types[4])(double) = {print_value, print_accumulated, print_square, print_abs};

    double num1;
    double num2;
    double num3;

    printf("Enter your numbers: ");
    error = scanf("%lf %lf %lf",&num1,&num2,&num3);
    if(error != 3)
    {
        printf("Incorrect input");
        return 1;
    }

    if(num2 > 0 && num1 > num3)
    {
        printf("Incorrect input data");
        return 2;
    }
    if(num1 < num3 && num2 < 0)
    {
        printf("Incorrect input data");
        return 2;
    }
    if(num2 == 0)
    {
        printf("Incorrect input data");
        return 2;
    }

    int type;

    printf("Enter your type: ");
    error = scanf("%i",&type);
    if(error != 1)
    {
        printf("Incorrect input");
        return 1;
    }
    if(type < 0 || type > 3)
    {
        printf("Incorrect input data");
        return 2;
    }

    for_loop(num1, num2, num3, *(types + type));
    return 0;
}

File: Dante2/Module2.07/Zadanie1.04/main.rs
use crate::for_loop::{for_loop, print_abs, print_accumulated, print_square, print_value};

mod for_loop;

fn main() {
    let types: &[fn(f64)] = &[print_abs, print_accumulated, print_square, print_value];

    let mut input: String = String::new();

    println!("Enter start: ");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let start: f64 = input.trim().parse().expect("Incorrect parse");

    println!("Enter step: ");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let step: f64 = input.trim().parse().expect("Incorrect parse");

    println!("Enter end: ");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let end: f64 = input.trim().parse().expect("Incorrect parse");

    if step > 0.0 && start > end || step == 0.0 {
        panic!("Incorrect input data");
    }
    if start < end && step < 0.0 {
        panic!("Incorrect input data");
    }

    println!("Enter type: ");
    input.clear();
    std::io::stdin().read_line(&mut input).expect("Incorrect input");
    let pick: usize = input.trim().parse().expect("Incorrect parse");

    if pick > 3 {
        panic!("Incorrect input");
    }

    for_loop(start, step, end, types[pick]);
}
